services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java

```diff
@@ -6509,6 +6509,48 @@ public boolean isLeaderControllerFor(String clusterName) {
     return model.isLeader();
   }
 
+  @Override
+  public InstanceRemovableStatuses getAggregatedHealthStatus(
+      String cluster,
+      Set<String> instances,
+      List<String> toBeStoppedInstances) {
+    InstanceRemovableStatuses statuses = new InstanceRemovableStatuses();
+
+    // If current controller is not the leader, redirect with leader URL
+    if (!isLeaderControllerFor(cluster)) {
+      Instance instance = getLeaderController(cluster);
+      statuses.setRedirectUrl(instance.getNodeId());
+      return statuses;
+    }
+
+    Map<String, String> nonStoppableInstances = new HashMap<>();
+    List<String> stoppableInstances = new ArrayList<>(toBeStoppedInstances);
+    statuses.setNonStoppableInstances(nonStoppableInstances);
+    statuses.setStoppableInstances(stoppableInstances);
+
+    // Check for maintenance, for ongoing maintenance none can be removed.
+    HelixVeniceClusterResources resources = getHelixVeniceClusterResources(cluster);
+    MaintenanceSignal maintenanceSignal = HelixUtils.getClusterMaintenanceSignal(cluster, resources.getHelixManager());
+
+    if (maintenanceSignal != null) {
+      for (String instanceId: instances) {
+        nonStoppableInstances.put(instanceId, InstanceRemovableStatuses.NonStoppableReason.ONGOING_MAINTENANCE.name());
+      }
+      return statuses;
+    }
```

Comment:
Why do we want to handle MM as a special case?

Reply:
Its from the doc where it mentions 

{
  "stoppableInstances": ["instance1", "instance2"],
  "nonStoppableInstancesWithReasons": {
    "instance3": "MIN_ACTIVE_REPLICA_VIOLATION",
    "instance4": "ONGOING_OPERATION"
  }
}

Reply:
I don't think this was intended for MM. I think we can leave the MM one out, or just check the min active replica. We can skip rebalance. Let's check with Helix team on the behavior during MM